### DISCUSSION TITLE
NoLeftOverPizza

### DISCUSSION THREAD

**Discussion by sparksp** [jfmengels/elm-review#88] [Rule ideas]
**What the rule should do:**

Forbid simple uses of right pizza (`|>`) that would not add any parens if removed.

**What problems does it solve:**

Pizza can make the place look untidy.

**Example of things the rule would report:**

```elm
    context |> rememberElmJsonProject project
```

**Example of things the rule would not report:**

```elm
    getContext model |> rememberElmJsonProject project
```

```elm
    context
        |> rememberElmJsonProject project
        |> rememberSomethingElse something
```

**When (not) to enable this rule:**

If you really like pizza then this rule might not be for you. Right pizza are often used to show a pipeline, but right pizza may also be used to change the focus of an expression - this rule may complain about such expressions.

**I am looking for:**

- A rule called "NoLeftOverPizza". :joy:
- Someone to implement it (I'm happy to mentor and review)

**Comment by lydell** [comment]
I would like to see this for _left_ pizza. `getName <| person` can be refactored to `getName person`.

**Comment by jfmengels** [comment]
@lydell [`elm-review-noleftpizza`](https://package.elm-lang.org/packages/truqu/elm-review-noleftpizza/latest/NoLeftPizza) already offers that with `NoLeftPizza.rule NoLeftPizza.Redundant`.

**Comment by lydell** [comment]
Thanks! Given the name – _no_ left pizza – I didn’t expect finding that there. But I guess it makes sense now.

**Comment by jfmengels** [comment]
It wasn't there in the first iteration ;)

Back on the subject, I agree with this rule, this is a small pet peeve of mine and I often refactor things to remove this way manually.
I think this could be auto-fixed, but I guess it can be a painful one to write. I know that some of my colleagues will be very annoyed at this kind of reports if they need to fix it manually.

**Comment by jfmengels** [comment]
Note that this should be easier to fix nowadays thanks to `withSourceCodeExtractor`

**Comment by lishaduck** [comment]
I think this can be marked as implemented in [SiriusStarr/elm-review-pipeline-styles](https://package.elm-lang.org/packages/SiriusStarr/elm-review-pipeline-styles/latest/).
